app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -76,12 +88,35 @@ def self.generate(opts = {})
 
         if opts[:ipxe]
           ipxe_source_file = File.join(Setting[:bootdisk_ipxe_dir], 'ipxe.lkrn')
-          raise(Foreman::Exception, N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
+          raise Foreman::Exception.new(N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
 
           FileUtils.cp(ipxe_source_file, File.join(wd, 'build', 'ipxe'))
           File.open(File.join(wd, 'build', 'script'), 'w') { |file| file.write(opts[:ipxe]) }
         end
 
+        if opts[:grub]
+          grub_path = "/var/lib/tftpboot/grub2/grubx64.efi"
```

Comment:
This is quite a scary method, maybe this can make it a little better.

```ruby
FileUtils.mkdir_p(File.join(wd, 'build', 'EFI', 'BOOT'))

grub_files = ['grubx64.efi', 'shimx64.efi']

copy_grub2_files(files: grub_files, to: wd)
```

and then

```ruby
def copy_grub2_files(files:, to:)
  files.each do |file|
    from = File.join('/var/lib/tftpboot/grub2/', file)
    raise Foreman::Exception.new(N_('Ensure %{path} exists by enabling TFTP smart proxy module.'), path: from) unless File.exist?(from)
    FileUtils.cp(from, File.join(to, 'build', 'EFI', 'BOOT', file))
  end
end

Not complete, but I guess you get the idea...

Reply:
Well, sometimes simple copy is more readable than complex looping that can be error prone. But I can do it, I don't like extra method tho. By the way `File.join('/var/lib/tftpboot/grub2/', file)` is actually only half-way there, but let's do that. I hate File.join use - Foreman will never run natively on Windows. If ever, WSL uses proper UNIX paths... :-)

Reply:
Okay, after another refactoring I have to say that few copies of lines are sometimes better and this is the case. I introduced a bug when I was modifying the code and had to debug it simply because it was much harder to read. I am putting it back it was.

Reply:
Ok, works for me. Can you cover the bug with a test case instead? ;-)